pytorch_lightning/core/optimizer.py

```diff
@@ -0,0 +1,121 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import tempfile
+import collections
+import copy
+import inspect
+import re
+import types
+from abc import ABC
+from argparse import Namespace
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union, Mapping
+from torch.optim.optimizer import Optimizer
+
+import torch
+from pytorch_lightning import _logger as log
+from pytorch_lightning.core.grads import GradInformation
+from pytorch_lightning.core.hooks import CheckpointHooks, DataHooks, ModelHooks
+from pytorch_lightning.core.memory import ModelSummary
+from pytorch_lightning.core.saving import ALLOWED_CONFIG_TYPES, PRIMITIVE_TYPES, ModelIO
+from pytorch_lightning.core.step_result import Result
+from pytorch_lightning.utilities import rank_zero_warn, AMPType
+from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
+from pytorch_lightning.utilities.xla_device_utils import XLADeviceUtils
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.parsing import (
+    AttributeDict,
+    collect_init_args,
+    get_init_args,
+)
+from pytorch_lightning.callbacks import Callback
+from torch import ScriptModule, Tensor
+from torch.nn import Module
+
+
+TPU_AVAILABLE = XLADeviceUtils.tpu_device_exists()
+
+if TPU_AVAILABLE:
+    import torch_xla.core.xla_model as xm
+
+
+def do_nothing_closure():
+    return
+
+
+class LightningOptimizer(Optimizer):
+
+    def __init__(self, trainer, optimizer: Optimizer = None, optimizer_idx: Optional[int] = None):
+        self._trainer = trainer
```

Comment:
if we need a trainer reference here, can we use `weakref` for this? The optimiser should not hold a hard reference on the trainer keeping it alive

Reply:
Not sure what you meant there ? How do you get a weakref ?

Reply:
https://docs.python.org/3/library/weakref.html#weakref.proxy